SofaKernel/framework/sofa/core/objectmodel/BaseObject.cpp

```diff
@@ -267,15 +267,6 @@ void BaseObject::releaseAspect(int aspect)
 
 void BaseObject::init()
 {
-
-
-	for(VecData::const_iterator iData = this->m_vecData.begin(); iData != this->m_vecData.end(); ++iData)
```

Comment:
So the object doesn't check himself it's Data?
How should we do if not going through a GUI ?

Reply:
The design where an overridden method must call some its parent function is error prone and is even considered an anti-pattern (https://en.wikipedia.org/wiki/Call_super).

I think that Guillaume ideas to fix this is to delegate the checking to an external SceneChecker. The SceneChecker can be added to any application and thus is not specific to a (G)UI but as it is pointed by Erik it is a bit strange that object self-consistency is checked by an external process. 

The "required" field in a Data allows to implement a default check but, in real components, there are much more checking conditions to implement (eg: data fields are conditionally needed when a third one is set to true/false). These specific cases are not covered by the "required" field and still need to be implemented in the init method. Because of that I think this PR is not removing the need to have an in-heriting class to call its parent's init method to insure that each Data added at any level in the class hierarchy is properly checked. 

Given how flexible is the current "init" approach is I think it will be hard to find a solution that fits all existing cases. 

In the following I'm dropping some ideas on using lambdas & functor to implement (and reuse) component specific data checking. 
```cpp
class DataIsRequired : public DataCheck
{
public: 
    bool operator(Data* d){
         return d->isSet();  
    }
}

/// somewhere in InitData(bool isRequired)
{
        if(isRequired())
             object->addDataCheck(DataIsRequired(data))
}

bool BaseObject::checkAllDataFields()
{
         for(Check c : m_allChecks)
         {
               if(! c.validate() ) 
                    return false; 
         }
         return true; 
}

BaseObject::init()
{
     checkAllDataFields(); 
}

MyParentClass::MyParentClass()
{
    addDataCheck( [this](){
                                   if( this->m_a.isSet() && this->m_a.getValue() < 1000){
                                             msg_error() << "Too big value";
                                             return false; 
                                   }
                                   return true; 
                             } );    
}

MyClass::MyClass()
{
    addDataCheck( [this](){
                                   if( this->m_a.isSet() && this->m_b.isSet )
                                   {
                                           msg_error(this) << "A  and B are mutually exclusive data fields... you cannot set both of them at the same time"; 
                                           return false; 
                                   }
                                   return true; 
                             } );    
}
```